File: src/data/data_validator.py
"""
Data Validator
Validates data quality and integrity
"""

import pandas as pd
import numpy as np
from typing import Dict, List, Tuple
from ..utils.config import Config
from ..utils.logger import setup_logger
from ..utils.helpers import validate_data_range, detect_outliers_iqr

logger = setup_logger()

class DataValidator:
    """Validates data quality and integrity"""
    
    def __init__(self, data: pd.DataFrame):
        self.data = data
        self.validation_results = {}
        
    def validate_all(self) -> Dict[str, any]:
        """
        Run all validation checks
        
        Returns:
            Dictionary of validation results
        """
        logger.info("Running data validation checks")
        
        results = {
            'missing_values': self.check_missing_values(),
            'data_ranges': self.check_data_ranges(),
            'temporal_continuity': self.check_temporal_continuity(),
            'outliers': self.detect_outliers(),
            'consistency': self.check_consistency(),
            'overall_valid': True
        }
        
        # Determine overall validity
        results['overall_valid'] = (
            not results['missing_values']['has_missing'] and
            results['data_ranges']['all_valid'] and
            results['temporal_continuity']['is_continuous']
        )
        
        self.validation_results = results
        return results
    
    def check_missing_values(self) -> Dict[str, any]:
        """
        Check for missing values
        
        Returns:
            Dictionary with missing value information
        """
        missing_counts = self.data[Config.REQUIRED_COLUMNS].isnull().sum()
        
        return {
            'has_missing': missing_counts.sum() > 0,
            'counts': missing_counts.to_dict(),
            'total_missing': int(missing_counts.sum())
        }
    
    def check_data_ranges(self) -> Dict[str, any]:
        """
        Check if data values are within valid ranges
        
        Returns:
            Dictionary with range validation results
        """
        results = {}
        all_valid = True
        
        for indicator in Config.REQUIRED_COLUMNS[1:]:
            is_valid = validate_data_range(self.data[indicator], 0, 100)
            results[indicator] = {
                'valid': bool(is_valid),
                'min': float(self.data[indicator].min()),
                'max': float(self.data[indicator].max())
            }
            if not is_valid:
                all_valid = False
        
        return {
            'all_valid': all_valid,
            'indicators': results
        }
    
    def check_temporal_continuity(self) -> Dict[str, any]:
        """
        Check for gaps in time series
        
        Returns:
            Dictionary with temporal continuity information
        """
        dates = pd.to_datetime(self.data['Date'])
        
        # Calculate expected number of quarters
        date_range = (dates.max() - dates.min()).days
        expected_quarters = (date_range // 90) + 1
        actual_quarters = len(dates)
        
        # Check for gaps
        date_diffs = dates.diff()[1:]
        expected_diff = pd.Timedelta(days=90)
        gaps = date_diffs[date_diffs > expected_diff * 1.5]
        
        return {
            'is_continuous': len(gaps) == 0,
            'expected_quarters': int(expected_quarters),
            'actual_quarters': int(actual_quarters),
            'gaps_found': len(gaps),
            'gap_locations': [str(self.data.loc[idx, 'Quarter']) for idx in gaps.index] if len(gaps) > 0 else []
        }
    
    def detect_outliers(self) -> Dict[str, any]:
        """
        Detect outliers in indicators
        
        Returns:
            Dictionary with outlier information
        """
        results = {}
        
        for indicator in Config.REQUIRED_COLUMNS[1:]:
            outliers = detect_outliers_iqr(self.data[indicator])
            outlier_indices = self.data[outliers].index.tolist()
            
            results[indicator] = {
                'count': int(outliers.sum()),
                'percentage': float(outliers.sum() / len(self.data) * 100),
                'quarters': [self.data.loc[idx, 'Quarter'] for idx in outlier_indices],
                'values': [float(self.data.loc[idx, indicator]) for idx in outlier_indices]
            }
        
        return results
    
    def check_consistency(self) -> Dict[str, any]:
        """
        Check mathematical consistency between indicators
        
        Returns:
            Dictionary with consistency check results
        """
        # Employment Rate + Unemployment Rate should approximately equal 100
        er_ur_sum = self.data['ER'] + self.data['UR']
        tolerance = 2.0  # Allow 2% tolerance
        
        inconsistent = abs(er_ur_sum - 100) > tolerance
        
        return {
            'er_ur_consistent': not inconsistent.any(),
            'inconsistent_count': int(inconsistent.sum()),
            'inconsistent_quarters': [
                self.data.loc[idx, 'Quarter'] 
                for idx in self.data[inconsistent].index
            ] if inconsistent.any() else []
        }
    
    def get_validation_report(self) -> str:
        """
        Generate human-readable validation report
        
        Returns:
            Formatted validation report string
        """
        if not self.validation_results:
            self.validate_all()
        
        report = []
        report.append("=" * 60)
        report.append("DATA VALIDATION REPORT")
        report.append("=" * 60)
        report.append("")
        
        # Missing values
        mv = self.validation_results['missing_values']
        report.append(f"Missing Values: {'✓ None' if not mv['has_missing'] else f'✗ {mv[\"total_missing\"]} found'}")
        
        # Data ranges
        dr = self.validation_results['data_ranges']
        report.append(f"Data Ranges: {'✓ All valid' if dr['all_valid'] else '✗ Invalid values found'}")
        
        # Temporal continuity
        tc = self.validation_results['temporal_continuity']
        report.append(f"Temporal Continuity: {'✓ Continuous' if tc['is_continuous'] else f'✗ {tc[\"gaps_found\"]} gaps found'}")
        
        # Outliers
        report.append("\nOutliers Detected:")
        for indicator, info in self.validation_results['outliers'].items():
            report.append(f"  {indicator}: {info['count']} ({info['percentage']:.1f}%)")
        
        # Consistency
        cons = self.validation_results['consistency']
        report.append(f"\nConsistency Check: {'✓ Passed' if cons['er_ur_consistent'] else '✗ Failed'}")
        
        # Overall
        report.append("")
        report.append("=" * 60)
        overall = "✓ VALID" if self.validation_results['overall_valid'] else "✗ ISSUES FOUND"
        report.append(f"Overall Status: {overall}")
        report.append("=" * 60)
        
        return "\n".join(report)

File: server/services/data-validator.js
/**
 * Data Validator Service
 * Validates data quality and integrity
 */

class DataValidator {
  constructor(data) {
    this.data = data
    this.validationResults = {}
  }

  /**
   * Run all validation checks
   * @returns {Object} Validation results
   */
  validateAll() {
    console.log("[DataValidator] Running validation checks")

    const results = {
      missingValues: this.checkMissingValues(),
      dataRanges: this.checkDataRanges(),
      temporalContinuity: this.checkTemporalContinuity(),
      outliers: this.detectOutliers(),
      consistency: this.checkConsistency(),
      overallValid: true,
    }

    results.overallValid =
      !results.missingValues.hasMissing && results.dataRanges.allValid && results.temporalContinuity.isContinuous

    this.validationResults = results
    return results
  }

  /**
   * Check for missing values
   * @returns {Object} Missing value information
   */
  checkMissingValues() {
    const indicators = ["LFPR", "ER", "UR", "UER"]
    const counts = {}
    let totalMissing = 0

    indicators.forEach((indicator) => {
      const missing = this.data.filter((record) => record[indicator] === null || record[indicator] === undefined).length
      counts[indicator] = missing
      totalMissing += missing
    })

    return {
      hasMissing: totalMissing > 0,
      counts,
      totalMissing,
    }
  }

  /**
   * Check if data values are within valid ranges
   * @returns {Object} Range validation results
   */
  checkDataRanges() {
    const indicators = ["LFPR", "ER", "UR", "UER"]
    const results = {}
    let allValid = true

    indicators.forEach((indicator) => {
      const values = this.data.map((r) => r[indicator]).filter((v) => v !== null && v !== undefined)

      const min = Math.min(...values)
      const max = Math.max(...values)
      const valid = min >= 0 && max <= 100

      results[indicator] = { valid, min, max }
      if (!valid) allValid = false
    })

    return { allValid, indicators: results }
  }

  /**
   * Check for gaps in time series
   * @returns {Object} Temporal continuity information
   */
  checkTemporalContinuity() {
    const dates = this.data.map((r) => new Date(r.Date))
    const gaps = []

    for (let i = 1; i < dates.length; i++) {
      const daysDiff = (dates[i] - dates[i - 1]) / (1000 * 60 * 60 * 24)
      if (daysDiff > 135) {
        // More than 4.5 months
        gaps.push({
          index: i,
          quarter: this.data[i].Quarter,
          daysSinceLastRecord: daysDiff,
        })
      }
    }

    return {
      isContinuous: gaps.length === 0,
      gapsFound: gaps.length,
      gaps,
    }
  }

  /**
   * Detect outliers using IQR method
   * @returns {Object} Outlier information
   */
  detectOutliers() {
    const indicators = ["LFPR", "ER", "UR", "UER"]
    const results = {}

    indicators.forEach((indicator) => {
      const values = this.data
        .map((r) => r[indicator])
        .filter((v) => v !== null && v !== undefined)
        .sort((a, b) => a - b)

      const q1 = values[Math.floor(values.length * 0.25)]
      const q3 = values[Math.floor(values.length * 0.75)]
      const iqr = q3 - q1
      const lowerBound = q1 - 1.5 * iqr
      const upperBound = q3 + 1.5 * iqr

      const outliers = this.data
        .map((record, idx) => ({ record, idx }))
        .filter(({ record }) => record[indicator] < lowerBound || record[indicator] > upperBound)

      results[indicator] = {
        count: outliers.length,
        percentage: (outliers.length / this.data.length) * 100,
        outliers: outliers.map(({ record, idx }) => ({
          quarter: record.Quarter,
          value: record[indicator],
          index: idx,
        })),
      }
    })

    return results
  }

  /**
   * Check mathematical consistency
   * @returns {Object} Consistency check results
   */
  checkConsistency() {
    const inconsistent = []
    const tolerance = 2.0

    this.data.forEach((record, idx) => {
      const sum = record.ER + record.UR
      if (Math.abs(sum - 100) > tolerance) {
        inconsistent.push({
          index: idx,
          quarter: record.Quarter,
          erUrSum: sum,
          difference: Math.abs(sum - 100),
        })
      }
    })

    return {
      erUrConsistent: inconsistent.length === 0,
      inconsistentCount: inconsistent.length,
      inconsistencies: inconsistent,
    }
  }

  /**
   * Generate validation report
   * @returns {string} Formatted report
   */
  getValidationReport() {
    if (Object.keys(this.validationResults).length === 0) {
      this.validateAll()
    }

    const r = this.validationResults
    const lines = [
      "=".repeat(60),
      "DATA VALIDATION REPORT",
      "=".repeat(60),
      "",
      `Missing Values: ${r.missingValues.hasMissing ? `✗ ${r.missingValues.totalMissing} found` : "✓ None"}`,
      `Data Ranges: ${r.dataRanges.allValid ? "✓ All valid" : "✗ Invalid values found"}`,
      `Temporal Continuity: ${r.temporalContinuity.isContinuous ? "✓ Continuous" : `✗ ${r.temporalContinuity.gapsFound} gaps found`}`,
      "",
      "Outliers Detected:",
      ...Object.entries(r.outliers).map(([ind, info]) => `  ${ind}: ${info.count} (${info.percentage.toFixed(1)}%)`),
      "",
      `Consistency Check: ${r.consistency.erUrConsistent ? "✓ Passed" : "✗ Failed"}`,
      "",
      "=".repeat(60),
      `Overall Status: ${r.overallValid ? "✓ VALID" : "✗ ISSUES FOUND"}`,
      "=".repeat(60),
    ]

    return lines.join("\n")
  }
}

module.exports = DataValidator
